pkg/guardian/jail.go

```diff
@@ -0,0 +1,165 @@
+	return &RouteJailProvider{store: store, logger: logger}
+}
+
+// routeJailKeyFunc provides a key unique to a particular client IP and request path.
+func routeJailKeyFunc(req Request) string {
+	return "jailer:" + req.RemoteAddress + ":" + req.Path
+}
+
+type GenericJailer struct {
+	keyFunc      func(req Request) string
+	jailProvider JailProvider
+	store        PrisonerStore
+	counter      Counter
+	logger       logrus.FieldLogger
+	onJailsHandled []JailsHandledHook
+}
+
+type JailsHandledHook func(req Request, blocked bool, dur time.Duration, err error)
+
+func OnGenericJailerHandled(mr MetricReporter) JailsHandledHook {
+	return func(req Request, blocked bool, dur time.Duration, err error) {
+		opts := []MetricOptionSetter{}
+		if blocked {
+			// Only set additional tags for requests that were blocked. This way, we know the cardinality of the custom metrics.
+			opts = append(opts, WithRemoteAddress(req.RemoteAddress))
+		}
+		mr.HandledJail(req, blocked, err != nil, dur, opts...)
+	}
+}
+
+func (gj *GenericJailer) IsBanned(ctx context.Context, request Request) (banned bool, err error) {
+	if gj.keyFunc == nil || gj.jailProvider == nil || gj.counter == nil || gj.logger == nil || gj.store == nil {
+		gj.logger.Error("misconfigured generic jailer: missing required field")
+		return false, nil
+	}
+
+	start := time.Now().UTC()
+	defer func() {
+		for _, fn := range gj.onJailsHandled {
+			fn(request, banned, time.Since(start), err)
+		}
+	}()
+
+	banned = gj.store.IsPrisoner(request.RemoteAddress)
+	if banned {
+		return banned, nil
+	}
+
+	jail := gj.jailProvider.GetJail(request)
+	gj.logger.Debugf("fetched jail %v for request %v", jail, request.Path)
+	if !jail.Limit.Enabled {
+		gj.logger.Debugf("jail limit not enabled for %v, allowing", request)
+		return false, nil
+	}
+
+	key := SlotKey(gj.keyFunc(request), time.Now().UTC(), jail.Limit.Duration)
+	gj.logger.Debugf("generated key %v for request %v", key, request)
+
+	currCount, blocked, err := gj.counter.Incr(ctx, key, 1, jail.Limit.Count, jail.Limit.Duration)
+	if err != nil {
+		err := errors.Wrap(err, fmt.Sprintf(" error incrementing counter for request: %v", request))
+		gj.logger.WithError(err).Errorf("error incrementing counter")
+		return false, err
+	}
+
+	banned = blocked || currCount > jail.Limit.Count
+	if banned {
+		gj.store.AddPrisoner(request.RemoteAddress, jail)
+		gj.logger.Debugf("banning ip: %v, due to jail: %v", request.RemoteAddress, jail)
+	}
+
+	return banned, nil
+}
+
+func NewGenericJailer(store RouteJailStore, logger logrus.FieldLogger, c Counter, s PrisonerStore, mr MetricReporter) *GenericJailer {
+	return &GenericJailer{
+		keyFunc:      routeJailKeyFunc,
+		jailProvider: NewRouteJailProvider(store, logger),
+		store:        s,
+		counter:      c,
+		logger:       logger,
+		// TODO: Expose this field for users to configure as needed
+		onJailsHandled: []JailsHandledHook{OnGenericJailerHandled(mr)},
```

Comment:
I'm not sure why this needs to be generic if it's only used for metrics. An unbounded array of arbitrary hooks in the request path makes me nervous for future misuse.

Reply:
that's fair. to be honest, I don't really think we should expose basically anything from Guardian as a consumable Go package. It has lots of implicit requirements to work properly that could be broken by users due to the amount we expose already. 

i will remove this comment and perhaps remove the concept of hooks here. I just think it's valuable to have some custom metric reporting depending on how the `GenericJailer` is configured, so I will think of a more safe way to do this.